clientlibrary/worker/worker.go

```diff
@@ -284,10 +310,23 @@ func (w *Worker) eventLoop() {
 				}
 
 				// The shard is closed and we have processed all records
-				if shard.Checkpoint == chk.SHARD_END {
+				if shard.GetCheckpoint() == chk.SHARD_END {
 					continue
 				}
 
+				var stealShard bool
+				if w.kclConfig.EnableLeaseStealing && shard.ClaimRequest != "" {
+					if shard.GetLeaseTimeout().Before(time.Now().Add(time.Duration(w.kclConfig.LeaseStealingIntervalMillis) * time.Millisecond)) {
+						if shard.ClaimRequest != w.workerID {
+							log.Debugf("Shard being stolen: %s", shard.ID)
+							continue
+						} else {
+							stealShard = true
+							log.Debugf("Stealing shard: %s", shard.ID)
+						}
+					}
+				}
+
 				err = w.checkpointer.GetLease(shard, w.workerID)
```

Comment:
I am thinking about why we cannot relinquish a shard or don't try to get lease more than optimal shards per worker even though it is below max shard numbers. Basically, I don't feel very comfortable on shard being stolen instead of be relinquished by its owner.

Reply:
I'm not sure how this would work predictably. If a worker is idle (has 0 shards) then it is not represented in the DynamoDB table. With that, workers cannot accurately calculate optimal shards per worker since they don't know how many workers with 0 shards there are. See https://github.com/connormckelvey/vmware-go-kcl/blob/293d7796ff5216d2f1a7102ea57351371004fe0c/clientlibrary/worker/worker.go#L422

Reply:
The dynamodb table has shardid as primary key. One approach is to create a special entry in table (fake shard) and put idle worker as attribute(s). A slightly different but easier implementation.

Reply:
Seems like it would be hard to:
1) Guarantee that relinquished shards get picked up by workers that need them and
2) If a worker gets into a bad state that some other worker picks up it shards to ensure every shard is being processed.

We desperately need lease stealing in our production deployment of Vmware Go KCL. To be honest, we are not willing to refactor this entire implementation. Totally open to fixing bugs/tests/small design decisions. But if we don't align on the fundamental approach to load balancing shards, we will need to move forward with our implementation as a fork of vmware go kcl and not contribute back up stream.

Reply:
I am quite hesitate to rush to solution especially for adding a relatively big feature. Please fork this repo and I will bring back those changes as needed. Thx!